### PR TITLE
Feat/ex 2

### PR DESCRIPTION
**Pull request by talivneh**
Hi Mentors,
this is what I have done so far.
Also, I added some bonuses (such as adding Pokemon by name as well as ID, alert if input is invalid, and option to add multiple items, from different types, by separating them with comma)

### CODE REVIEW

File: src/ex2/.vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+    // Use IntelliSense to learn about possible attributes.
```
Comment (monadav) [inline src/ex2/.vscode/launch.json]: actually you dont have to add these to the repo as this is not a source code but an IDE configuration file, you can add it to the `.gitingore` file :)

File: src/ex2/ItemManager.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+export default class ItemManager {
+  constructor() {
+    this.itemsList = [];
+    this.pokemonClient = new PokemonClient();
+  }
+
+  async addItem(item) {
+    const isPokemonFormat = /^[0-9]*$/;
+    if (isPokemonFormat.test(item.text)) {
```
Comment (monadav) [inline src/ex2/ItemManager.js]: would be nice to extract it to `isValidPokemonId` method in pokemonClient

File: src/ex2/PokemonClient.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+      return pokemonsJson.results;
+    } catch {
+      (err) => console.log(err);
+    }
+  }
+
+  async fetchPokemonNameById(id) {
+    try {
+      const pokemonsList = await this.fetchAllPokemons();
```
Comment (monadav) [inline src/ex2/PokemonClient.js]: no need to fetch all pokemons here!
`https://pokeapi.co/api/v2/pokemon/1` will fetch pokemon with ID 1,
same for ` https://pokeapi.co/api/v2/pokemon/pikachu` to fetch pokemon by ID

File: src/ex2/PokemonClient.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      if (pokemonsList[id].name) {
```
Comment (monadav) [inline src/ex2/PokemonClient.js]: the list we get has offset of 1 index, first pokemon (id 1) is in first index (0) so there's a bug here :)

File: src/ex2/PokemonClient.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        return `Catch ${pokemonsList[id].name}`;
```
Comment (monadav) [inline src/ex2/PokemonClient.js]: this method should be generic and not return full sentence (Catch xxx), it has 1 purpose which to fetch the desired pokemon and return its name

File: src/ex2/PokemonClient.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+      } else {
+        return `Faild to fetch Pokemon with ID ${id}`;
```
Comment (monadav) [inline src/ex2/PokemonClient.js]: better throw an error here and handle it as you would like

File: src/ex2/View.js [hunk 7]
```diff
@@ -0,0 +1,196 @@
+export default class View {
+  constructor(addItemHandler, deleteItemHandler, getAllItemsHandler) {
+    this.toDoInput = document.querySelector("input");
+    this.listContainer = document.querySelector("ul");
+    this.alertBox = document.querySelector(".alert");
+    this.alertBoxText = document.querySelector(".alert-innet-text");
+    this.addItemHandler = addItemHandler;
```
Comment (monadav) [inline src/ex2/View.js]: instead of passing props you could have pass an entire instance of `ItemManager`

File: src/ex2/View.js [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+  get _inputText() {
+    return this.toDoInput.value;
+  }
+
+  resetInput() {
+    this.toDoInput.value = "";
+  }
+
+  getElement(selector) {
```
Comment (monadav) [inline src/ex2/View.js]: `getElementBySelector`, the function name must be understandable!

File: src/ex2/View.js [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    const element = document.querySelector(selector);
+    return element;
+  }
+
+  handleAddBtnClick() {
+    if (this._inputText) {
```
Comment (monadav) [inline src/ex2/View.js]: when pressing on spacebar (input = ' ') it will enter this IF section

File: src/ex2/View.js [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+  handleInput(text) {
+    const textList = text.split(",");
+    if (textList.length > 1) {
+      textList.forEach((text) => this.validateInput(text.trim()));
+    } else {
+      this.validateInput(text.trim());
+    }
+  }
+
+  validateInput(text) {
```
Comment (monadav) [inline src/ex2/View.js:62]: input validation should only validate the input, not handle it further.
better return true/false (if it is valid or not)

File: src/ex2/View.js [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+    const itemsList = this.getAllItemsHandler();
+    const isExists = itemsList.some((item) => item.text === text);
+
+    if (isExists) {
```
Comment (monadav) [inline src/ex2/View.js:74]: for some reason I was able to add duplicated items here, maybe a bug 🐛

File: src/ex2/View.js [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+  }
+
+  addNewItem(text) {
+    const time = new Date().toLocaleDateString();
+    const item = {
+      text,
+      time,
+      complete: false,
+      checkTime: null,
+      toggleComplete() {
```
Comment (monadav) [inline src/ex2/View.js:114]: toggle complete function is not working, the pending number is not changing when marking as done

**Review by monadav — COMMENTED** [review]
Hey! great code here!
Left some comments, please look at them.
Also, I've found some bugs regarding the synchronization of your item list, one of the requirements was to `Render todos from the array using a separate render method` which mean to have one render function that will re-render all the items in the list upon each change (add, remove, complete and more..)
the function should iterate the items array and render them over and over again :) try to make it happen! 🥳

File: src/ex2/PokemonClient.js (hunk 3 above)
Comment (talivneh) [reply src/ex2/PokemonClient.js]: Hi, don't know if you saw my comment there but I wrote that I found out that fetching all Pokemons and then query them by ID is a lot faster then trying to fetch each of them by id. I will change it anyway :)

File: src/ex2/PokemonClient.js (hunk 4 above)
Comment (talivneh) [reply src/ex2/PokemonClient.js:22]: fixed it :)

Comment (talivneh) [reply src/ex2/PokemonClient.js]: fixed with previous changes

File: src/ex2/PokemonClient.js (hunk 5 above)
Comment (talivneh) [reply src/ex2/PokemonClient.js]: fixed :)

File: src/ex2/View.js (hunk 7 above)
Comment (talivneh) [reply src/ex2/View.js]: I tried to refactor but It keeps on causing errors :\ 
I put those on the "Main.js" constructor: 
image
and this on the View.js:
image

Can you tell what am I doing wrong? 

I'm getting "undefined" when trying to get values from the itemManager methods inside the view class

File: src/ex2/View.js (hunk 8 above)
Comment (talivneh) [reply src/ex2/View.js]: changed 👍

File: src/ex2/View.js (hunk 9 above)
Comment (talivneh) [reply src/ex2/View.js]: 👍  added trim()

File: src/ex2/View.js (hunk 10 above)
Comment (talivneh) [reply src/ex2/View.js:62]: 👍 made the necessary changes

File: src/ex2/View.js (hunk 11 above)
Comment (talivneh) [reply src/ex2/View.js:74]: fixed it :)

File: src/ex2/View.js (hunk 12 above)
Comment (talivneh) [reply src/ex2/View.js:114]: I added the rendering method (you specified in the top of this PR) and then fixed this issue as well :)

**Comment by talivneh** [comment]
Hi Nadav, 
Thanks for the detailed feedback! 
I made all the changes except for one that I tried to fixed and faced some issues-
all specified in the relevant comment.

Also, I notice that trying to fetch Pokemons by name causing everything to be really slow
(the reason is the program checks each and every text input to see if it's a valid Pokemon name, and this process takes time...)
Is there anything I can do to make it better?

Thanks again for the feedback, and waiting for your answer.

**Pull request merged**